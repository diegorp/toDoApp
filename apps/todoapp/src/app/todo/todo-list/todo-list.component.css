.flex-wrapper {
    display: flex;
    flex-direction: column;
    margin: 12px 0;
}

todoapp-todo-item {
    margin-bottom: 6px;
}

p {
    font-size: 3.5em;
    color: #dfdfdf;
    text-align: center;
    margin: 50px;
}