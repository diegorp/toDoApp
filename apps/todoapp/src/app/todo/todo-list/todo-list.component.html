<div class="flex-wrapper">
    <ng-container *ngIf="toDoList?.length; else emptyList"></ng-container>
    <ng-container *ngFor="let toDo of toDoList">
        <todoapp-todo-item [toDo]="toDo" (completeToDo)="onCompleteToDo(toDo)"></todoapp-todo-item>
    </ng-container>
</div>

<ng-template #emptyList>
    <p>Good job!</p>
    <p>You have no pending stuff to do!</p>
</ng-template>