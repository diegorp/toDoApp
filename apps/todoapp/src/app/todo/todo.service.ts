import { Injectable } from '@angular/core';
import { Observable, Subject } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class ToDoService {
  
  private readonly toDo = new Subject<string>();

  getToDo(): Observable<string> {
    return this.toDo.asObservable();
  }

  addToDo(newToDo: string) {
    this.toDo.next(newToDo);
  }
}
