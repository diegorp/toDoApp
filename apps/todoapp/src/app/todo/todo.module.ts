import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TodoItemComponent } from './todo-item/todo-item.component';
import { TodoListComponent } from './todo-list/todo-list.component';
import { TodoBuilderComponent } from './todo-builder/todo-builder.component';



@NgModule({
  declarations: [
    TodoItemComponent,
    TodoListComponent,
    TodoBuilderComponent
  ],
  imports: [
    CommonModule
  ],
  exports: [
    TodoListComponent,
    TodoBuilderComponent
  ]
})
export class TodoModule { }
