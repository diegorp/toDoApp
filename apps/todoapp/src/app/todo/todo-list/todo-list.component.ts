import { Component, OnDestroy, OnInit } from '@angular/core';
import { Subscription } from 'rxjs';
import { ToDoService } from '../todo.service';

@Component({
  selector: 'todoapp-todo-list',
  templateUrl: './todo-list.component.html',
  styleUrls: ['./todo-list.component.css']
})
export class TodoListComponent implements OnInit, OnDestroy {

  toDoList: string[] = [];
  
  private toDoSubscription: Subscription | undefined;

  constructor(private readonly toDoService: ToDoService) {}

  ngOnInit(): void {
      this.toDoSubscription = this.toDoService.getToDo()
        .subscribe(toDo => this.toDoList.push(toDo));
  }

  ngOnDestroy(): void {
      if (this.toDoSubscription) {
        this.toDoSubscription.unsubscribe();
      }
  }

  onCompleteToDo(toDo: string) {
    this.toDoList = this.removeToDoFromList(toDo);
  }

  private removeToDoFromList(toDo: string) {
    return this.toDoList.filter(toDoListItem => toDoListItem !== toDo);
  }
}
