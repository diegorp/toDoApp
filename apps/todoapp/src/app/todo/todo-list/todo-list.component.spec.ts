import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Subject } from 'rxjs';
import { ToDoService } from '../todo.service';

import { TodoListComponent } from './todo-list.component';

jest.mock('../todo.service');

describe('TodoListComponent', () => {
  let component: TodoListComponent;
  let fixture: ComponentFixture<TodoListComponent>;
  let toDoService: ToDoService;

  const toDoSubjectMock = new Subject<string>();

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ TodoListComponent ],
      providers: [ ToDoService ],
      schemas: [ NO_ERRORS_SCHEMA ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(TodoListComponent);
    component = fixture.componentInstance;
    toDoService = TestBed.inject(ToDoService);
    jest.spyOn(toDoService, 'getToDo').mockReturnValue(toDoSubjectMock.asObservable());

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should add a to do', () => {
    toDoListLengthIs(0);
    toDoSubjectMock.next('To Do 1');
    toDoListLengthIs(1);
    toDoSubjectMock.next('To Do 2');
    toDoListLengthIs(2);
  });

  it('should remove a to do', () => {
    toDoListLengthIs(0);
    toDoSubjectMock.next('To Do 1');
    toDoListLengthIs(1);
    fixture.detectChanges();
    component.onCompleteToDo('To Do 1');
    toDoListLengthIs(0);
  });

  it('should render an empty list', () => {
    toDoListLengthIs(0);
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render a list', () => {
    toDoSubjectMock.next('To Do 1');
    toDoSubjectMock.next('To Do 2');
    fixture.detectChanges();
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  function toDoListLengthIs(expectedLength: number) {
    expect(component.toDoList.length).toEqual(expectedLength);
  }
});
