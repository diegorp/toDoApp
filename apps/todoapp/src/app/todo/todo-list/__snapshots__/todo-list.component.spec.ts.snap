// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListComponent should render a list 1`] = `
<div
  id="root4"
>
  <div
    class="flex-wrapper"
  >
    
    
    <todoapp-todo-item />
    
    <todoapp-todo-item />
    
    
  </div>
  
</div>
`;

exports[`TodoListComponent should render an empty list 1`] = `
<div
  id="root3"
>
  <div
    class="flex-wrapper"
  >
    <p>
      Good job!
    </p>
    <p>
      You have no pending stuff to do!
    </p>
    
    
  </div>
  
</div>
`;
